clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```
@@ -0,0 +1,914 @@
+                    /*
+                     We have a network request in progress. We record the time of the request
+                     submission, so wait that amount of the time PLUS the requestTimeout that
+                     is provided.
+                    */
+                    String apiName = (localState == ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS) ? ApiKeys.GET_TELEMETRY_SUBSCRIPTIONS.name : ApiKeys.PUSH_TELEMETRY.name;
+                    timeMs = requestTimeoutMs;
+                    msg = String.format("the remaining wait time for the %s network API request, as specified by %s", apiName, CommonClientConfigs.REQUEST_TIMEOUT_MS_CONFIG);
+                    break;
+                }
+
+                case TERMINATING_PUSH_NEEDED: {
+                    timeMs = 0;
+                    msg = String.format("the client should try to submit the final %s network API request ASAP before closing", ApiKeys.PUSH_TELEMETRY.name);
+                    break;
+                }
+
+                case SUBSCRIPTION_NEEDED:
+                case PUSH_NEEDED: {
+                    String apiName = (localState == ClientTelemetryState.SUBSCRIPTION_NEEDED) ? ApiKeys.GET_TELEMETRY_SUBSCRIPTIONS.name : ApiKeys.PUSH_TELEMETRY.name;
+                    long timeRemainingBeforeRequest = localLastRequestMs + localIntervalMs - now;
+                    if (timeRemainingBeforeRequest <= 0) {
+                        timeMs = 0;
+                        msg = String.format("the wait time before submitting the next %s network API request has elapsed", apiName);
+                    } else {
+                        timeMs = timeRemainingBeforeRequest;
+                        msg = String.format("the client will wait before submitting the next %s network API request", apiName);
+                    }
+                    break;
+                }
+                default:
+                    // Should never get to here
```

If it would indicate a bug, should we rather throw an `InvalidStateException()` to surface the bug instead of masking it?

---

Done.